README.rst
contours - contour calculation with matplotlib
---------------------------------------------------------

.. begin-description

The `contours` package exposes Matplotlib's contour generators to the user
providing an equivalent of MATLAB's ``contourc`` function.  This is done for
grids (including curvilinear) as well as unstructured data via Delaunay
triangulation (**FUTURE**).

Regular contours can be returned as `NumPy`_ arrays or as `Shapely`_
LineStrings and LinearRings.  Filled contours can be returned as `NumPy`_
arrays or as `Shapely`_ Polygons.

Contour calculation without plotting is currently an open issue `#367`_ for
Matplotlib.  In that discussion, the type of hackery done in this library is
discouraged by the Matplotlib developers.  As of this writing there has been no
decision as to whether exposing contour calculations is a feature that should
be included in Matplotlib or not.  If such a refactoring in Matplotlib ever
takes place, this library will be refactored to avoid using Matplotlib's
internal components.

.. _Shapely: http://toblerity.org/shapely/manual.html

.. _NumPy: http://www.numpy.org

.. _#367: https://github.com/matplotlib/matplotlib/issues/367

.. end-description


.. begin-body

.. begin-links

Links
-----

* `Download <https://github.com/ccarocean/python-contours/archive/master.zip>`_
* `GitHub <https://github.com/ccarocean/python-contours>`_
* `Documentation <https://ccarocean.github.io/python-contours>`_

.. end-links

.. begin-features

Features
--------

* Filled and non-filled contours.
* Contours on regular grids.
* **TODO** Contours on unstructured data.

.. end-features


Examples
--------

A contrived example using `QuadContourGenerator` to compute the area and
circumference of a circle and a ring.

>>> from contours.core import shapely_formatter as shapely_fmt
>>> from contours.quad import QuadContourGenerator
>>> import numpy as np
>>> x = np.arange(-1, 1+0.01, 0.01)
>>> y = np.arange(-1, 1+0.01, 0.01)
>>> z = np.sqrt(x[:, np.newaxis]**2 + y[np.newaxis, :]**2)
>>> c = QuadContourGenerator.from_rectilinear(x, y, z, shapely_fmt)
>>> contour = c.filled_contour(max=1.0)
>>> print('Area: {:0.2f}'.format(contour[0].area))
Area: 3.14
>>> print('Length: {:0.2f}'.format(contour[0].length))
Length: 6.28
>>> contour = c.filled_contour(min=0.5, max=1.0)
>>> print('Area: {:0.2f}'.format(contour[0].area))
Area: 2.36
>>> print('Length: {:0.2f}'.format(contour[0].exterior.length))
Length: 6.28
>>> print('Length: {:0.2f}'.format(contour[0].interiors[0].length))
Length: 3.14


.. begin-author

Author
------

The `contours` module was written by Michael R. Shannon (`@mrshannon
<https://github.com/mrshannon>`_) <mrshannon.aerospace@gmail.com> in 2017.

It is maintained by:

* Michael R. Shannon (`@mrshannon <https://github.com/mrshannon>`_)
  <mrshannon.aerospace@gmail.com> 2017-

.. end-author

Testing
-------

**TODO**

.. end-body


CHANGES.rst
Changelog (contours)
====================

0.0.2
-----
* Bugfix: Add support for degenerate (single point) polygons in the
  shapely_formatter.

0.0.1
-----

* Curvilinear, rectilinear, and uniform grid contours.
* New output format: List of NumPy arrays.
* New output format: Single NumPy array (MATLAB style).
* New output format: List of Shapely geometry elements.


searchindex.js
Search.setIndex({docnames:["about","examples/index","handbook/index","index","installation","reference/core","reference/index","reference/quad","reference/summary","releasenotes/0.0.1","releasenotes/0.0.2","releasenotes/index"],envversion:53,filenames:["about.rst","examples/index.rst","handbook/index.rst","index.rst","installation.rst","reference/core.rst","reference/index.rst","reference/quad.rst","reference/summary.rst","releasenotes/0.0.1.rst","releasenotes/0.0.2.rst","releasenotes/index.rst"],objects:{"":{contours:[6,0,0,"-"]},"contours.core":{ContourMixin:[5,1,1,""],MPLPATHCODE:[5,1,1,""],matlab_formatter:[5,4,1,""],numpy_formatter:[5,4,1,""],sample_formatter:[5,4,1,""],shapely_formatter:[5,4,1,""]},"contours.core.ContourMixin":{contour:[5,2,1,""],filled_contour:[5,2,1,""],formatter:[5,3,1,""]},"contours.core.MPLPATHCODE":{CLOSEPOLY:[5,3,1,""],CURVE3:[5,3,1,""],CURVE4:[5,3,1,""],LINETO:[5,3,1,""],MOVETO:[5,3,1,""],STOP:[5,3,1,""]},"contours.quad":{QuadContourGenerator:[7,1,1,""]},"contours.quad.QuadContourGenerator":{contour:[7,2,1,""],filled_contour:[7,2,1,""],formatter:[7,3,1,""],from_curvilinear:[7,5,1,""],from_rectilinear:[7,5,1,""],from_uniform:[7,5,1,""]},contours:{core:[5,0,0,"-"],quad:[7,0,0,"-"]}},objnames:{"0":["py","module","Python module"],"1":["py","class","Python class"],"2":["py","method","Python method"],"3":["py","attribute","Python attribute"],"4":["py","function","Python function"],"5":["py","classmethod","Python class method"]},objtypes:{"0":"py:module","1":"py:class","2":"py:method","3":"py:attribute","4":"py:function","5":"py:classmethod"},terms:{"case":10,"class":9,"default":[5,7],"enum":5,"export":3,"function":[0,3,5,7],"import":[3,7],"public":5,"return":[0,3,5,7,9,10],"while":5,Axes:7,For:5,The:[0,3,5,7,9,10],There:5,These:5,Use:5,about:3,acw:5,aerospac:[0,3],algorithm:9,alia:7,also:5,alwai:5,ani:5,anoth:5,arang:[3,7],area:[3,7],arg:5,argument:[5,7],around:7,arrai:[0,3,5,7,9],array_lik:7,avoid:[0,3],axes:7,base:[5,7],basic:3,been:[0,3],begin:5,belong:5,below:5,between:[5,7],bezier:5,calcul:[0,3,5,6,7],call:[5,7],callabl:[5,7],can:[0,3,5,9],cartesian:7,ccw:5,circl:[3,7],circumfer:[3,7],classmethod:7,closepoli:5,code:5,column:[5,7],com:[0,3],common:5,compon:[0,3],comput:[3,7],consid:7,construct:[5,7],constructor:7,contain:5,contiain:5,contour:[0,4,5,6,7,10],contourc:[0,3],contourmixin:[5,7],contriv:[3,7],control:5,convers:[5,7],convert:[5,7],coordin:[5,7],copi:5,core:[3,6,7,10],corespond:5,could:10,cubic:5,current:[0,3,5],curv:5,curve3:5,curve4:5,curvilinear:[0,3,7,9],data:[0,3,5,7],decis:[0,3],defin:10,definit:7,delaunai:[0,3],depend:5,descart:5,descript:5,develop:[0,3],dimension:7,direct:5,discourag:[0,3],discuss:[0,3],distanc:7,document:[0,5],doe:5,done:[0,3],download:[0,3],draw:5,each:[5,7],effici:7,element:[5,9],enabl:5,end:5,endpoint:5,entir:5,equival:[0,3,7],ever:[0,3],exampl:7,explan:5,expos:[0,3],exterior:[3,5,7],extern:[5,7],fastest:5,fill:[0,3,5,7,10],filled_contour:[3,5,7],finfo:[5,7],first:5,float64:[5,7],follow:5,format:[3,5,7],formatt:7,from:[3,5,7],from_curvilinear:7,from_rectilinear:[3,7],from_uniform:7,further:5,futur:[0,3],gener:[0,3,5,7,9],geo:7,geojson:5,geometr:5,geometri:[5,9,10],get:[5,7],github:[0,3],give:5,given:[5,7],gmail:[0,3],gov:7,grid:[0,3,7],hackeri:[0,3],handbook:3,handl:10,has:[0,3],have:5,header:5,highli:7,hole:[5,10],http:7,ide:7,ignor:5,implement:5,includ:[0,3,5],index:[3,7],inform:[5,7],inherit:7,initi:7,insid:5,instal:3,instead:5,interfac:9,interior:[3,5,7],intern:[0,3,5,7],issu:[0,3],juli:7,kwarg:5,last:5,len:7,length:[3,7],level:[5,7],librari:[0,3,5],like:5,line:[5,7,10],linear:[0,3,5],linestr:[0,3,5],lineto:5,list:[5,9],lower:5,mai:5,maintain:[0,3],major:5,marker:5,mask:7,mathwork:5,matlab:[0,3,5,9],matlab_formatt:5,matplotlib:[0,3,5,7,9],matter:5,max:[3,5,7],maximum:[5,7],meant:5,method:[5,7],michael:[0,3],min:[3,5,7],minimum:[5,7],mixin:5,modul:[0,3,6],more:[5,7],move:5,moveto:5,mplpathcod:5,mrshannon:[0,3],multipl:5,must:7,ndarrai:5,newaxi:[3,7],noaa:7,non:[0,3],none:[5,7],note:[3,5],now:10,number:[5,7],numpi:[0,3,5,7,9],numpy_formatt:[5,7],nx2:5,object:[5,10],one:[5,7],onli:[5,7],open:[0,3],origin:7,other:7,packag:[0,3,5],page:3,pair:5,paramet:[5,7],patch:5,path:5,pen:5,php:7,pick:5,pip:4,place:[0,3],plot:[0,3],point:[5,7,10],polygon:[0,3,5,7],polylin:5,posit:5,possibl:[5,7],print:[3,7],provid:[0,3,5,9],pure:7,purpos:5,quad:[3,6,9],quadcontourgener:[3,7,9],quadrat:5,rectilinear:[7,9],refactor:[0,3],refer:[3,7],regular:[0,3,10],releas:3,remov:10,repres:5,requir:5,rest:5,result:[5,7],retriev:7,ring:[3,7],row:[5,7],rule:5,same:7,sample_formatt:5,search:3,second:5,see:[5,7],segment:5,shannon:[0,3],shape:[0,3,4,5,7,9],shapely_fmt:[3,7],shapely_formatt:[3,5,7,10],should:[0,3,5],sinc:5,singl:[5,9,10],size:7,some:5,sourc:[5,7],speedup:5,sqrt:[3,7],start:5,step:7,stop:5,structur:7,structured_grid:7,style:[5,9],summari:[3,6],support:[4,7],take:[0,3],thei:10,thi:[0,3,5,7],titl:7,todo:[0,1,2,3],transpos:5,triangul:[0,3],tupl:5,two:5,type:[0,3,5,7],uniform:[7,9],unstructur:[0,3],used:[5,7],user:[0,3],using:[0,3,7],valu:7,verbatim:5,vertex:5,vertic:5,via:[0,3],wai:5,well:[0,3,10],when:5,where:5,whether:[0,3],wiki:7,without:[0,3],wrapper:7,write:[0,3],written:[0,3,7]},titles:["About","Examples","Handbook","Welcome to contours\u2019s documentation!","Instalation","<code class=\"docutils literal\"><span class=\"pre\">core</span></code> Module","Reference","<code class=\"docutils literal\"><span class=\"pre\">quad</span></code> Module","Summary","0.0.1","0.0.2","Release Notes"],titleterms:{"class":[5,7],"export":6,about:0,author:[0,3],basic:4,bugfix:10,contour:[3,9],core:[5,8],degener:10,document:3,enumer:5,exampl:[1,3],featur:[0,3],format:9,formatt:[5,10],grid:9,handbook:2,indic:3,instal:4,link:[0,3],modul:[5,7,8],note:11,output:9,own:5,polygon:10,quad:[7,8],refer:6,releas:11,shape:10,summari:[5,7,8],tabl:3,test:3,welcom:3,writ:5,your:5}})

contours/core.py
# -* coding: utf-8 -*-
"""Common enums, functions, and classes for the `contours` package."""

# Python 2 support
# pylint: disable=redefined-builtin,unused-wildcard-import,wildcard-import
from __future__ import (absolute_import, division,
                        print_function, unicode_literals)
from builtins import *

import numbers
import itertools
from enum import IntEnum
import numpy as np

# Attempt to import shapely and enable speedups.
try:
    import shapely.speedups
    if shapely.speedups.available:
        shapely.speedups.enable()
finally:
    try:
        from shapely.geometry import LineString, LinearRing, Polygon
    except ImportError:
        pass


class MPLPATHCODE(IntEnum):
    """Matplotlib path codes.

    These are included in this library for the purpose of writting your own
    formatter.  The description of each one is copied verbatim from the
    documentation for :class:`matplotlib.path.Path`.
    """

    STOP = 0
    """1 vertex (ignored)

    A marker fo the end of the entire path (currently not required and
    ignored).
    """

    MOVETO = 1
    """1 vertex

    Pick up the pen and move to the given vertex.
    """

    LINETO = 2
    """1 vertex

    Draw a line from the current position to the given vertex.
    """

    CURVE3 = 3
    """1 control point, 1 endpoint

    Draw a quadratic Bezier curve from he current position, with the given
    control point, to the given end point.
    """

    CURVE4 = 4
    """2 control points, 1 endpoint

    Draw a cubic Bezier curve from the current position, with the given control
    points, to the given and point.
    """

    CLOSEPOLY = 79
    """1 vertex (ignored)

    Draw a line segment to the start point of the current polyline.
    """


def null_formatter(level, vertices, codes=None):
    """Null formatter that passes through the raw vertices and codes."""
    return level, vertices, codes


def numpy_formatter(_, vertices, codes=None):
    """`NumPy`_ style contour formatter.

    Contours are returned as a list of Nx2 arrays containing the x and y
    vertices of the contour line.

    For filled contours the direction of vertices matters:

    * CCW (ACW): The vertices give the exterior of a contour polygon.
    * CW: The vertices give a hole of a contour polygon.  This hole will
        always be inside the exterior of the last contour exterior.

    .. note:: This is the fastest format.

    .. _NumPy: http://www.numpy.org

    """
    if codes is None:
        return vertices
    numpy_vertices = []
    for vertices_, codes_ in zip(vertices, codes):
        starts = np.nonzero(codes_ == MPLPATHCODE.MOVETO)[0]
        stops = np.nonzero(codes_ == MPLPATHCODE.CLOSEPOLY)[0]
        for start, stop in zip(starts, stops):
            numpy_vertices.append(vertices_[start:stop+1, :])
    return numpy_vertices


def matlab_formatter(level, vertices, codes=None):
    """`MATLAB`_ style contour formatter.

    Contours are returned as a single Nx2, `MATLAB`_ style, contour array.
    There are two types of rows in this format:

    * Header: The first element of a header row is the level of the contour
      (the lower level for filled contours) and the second element is the
      number of vertices (to follow) belonging to this contour line.
    * Vertex: x,y coordinate pairs of the vertex.

    A header row is always followed by the coresponding number of vertices.
    Another header row may follow if there are more contour lines.

    For filled contours the direction of vertices matters:

    * CCW (ACW): The vertices give the exterior of a contour polygon.
    * CW: The vertices give a hole of a contour polygon.  This hole will
        always be inside the exterior of the last contour exterior.

    For further explanation of this format see the `Mathworks documentation
    <https://www.mathworks.com/help/matlab/ref/contour-properties.html#prop_ContourMatrix>`_
    noting that the MATLAB format used in the `contours` package is the
    transpose of that used by `MATLAB`_ (since `MATLAB`_ is column-major
    and `NumPy`_ is row-major by default).

    .. _NumPy: http://www.numpy.org

    .. _MATLAB: https://www.mathworks.com/products/matlab.html

    """
    vertices = numpy_formatter(level, vertices, codes)
    if codes is not None:
        level = level[0]
    headers = np.vstack((
        [v.shape[0] for v in vertices],
        [level]*len(vertices))).T
    vertices = np.vstack(
        list(it.__next__() for it in
             itertools.cycle((iter(headers), iter(vertices)))))
    return vertices


def shapely_formatter(_, vertices, codes=None):
    """`Shapely`_ style contour formatter.

    Contours are returned as a list of :class:`shapely.geometry.LineString`,
    :class:`shapely.geometry.LinearRing`, and :class:`shapely.geometry.Point`
    geometry elements.

    Filled contours return a list of :class:`shapely.geometry.Polygon`
    elements instead.

    .. note:: If possible, `Shapely speedups`_ will be enabled.

    .. _Shapely: http://toblerity.org/shapely/manual.html

    .. _Shapely speedups: http://toblerity.org/shapely/manual.html#performance


    See Also
    --------
    `descartes <https://bitbucket.org/sgillies/descartes/>`_ : Use `Shapely`_
    or GeoJSON-like geometric objects as matplotlib paths and patches.

    """
    elements = []
    if codes is None:
        for vertices_ in vertices:
            if np.all(vertices_[0, :] == vertices_[-1, :]):
                # Contour is single point.
                if len(vertices) < 3:
                    elements.append(Point(vertices_[0, :]))
                # Contour is closed.
                else:
                    elements.append(LinearRing(vertices_))
            # Contour is open.
            else:
                elements.append(LineString(vertices_))
    else:
        for vertices_, codes_ in zip(vertices, codes):
            starts = np.nonzero(codes_ == MPLPATHCODE.MOVETO)[0]
            stops = np.nonzero(codes_ == MPLPATHCODE.CLOSEPOLY)[0]
            try:
                rings = [LinearRing(vertices_[start:stop+1, :])
                        for start, stop in zip(starts, stops)]
                elements.append(Polygon(rings[0], rings[1:]))
            except ValueError as err:
                # Verify error is from degenerate (single point) polygon.
                if np.any(stop - start - 1 == 0):
                    # Polygon is single point, remove the polygon.
                    if stops[0] < starts[0]+2:
                        pass
                    # Polygon has single point hole, remove the hole.
                    else:
                        rings = [
                            LinearRing(vertices_[start:stop+1, :])
                            for start, stop in zip(starts, stops)
                            if stop >= start+2]
                        elements.append(Polygon(rings[0], rings[1:]))
                else:
                    raise(err)
    return elements


class ContourMixin(object):
    """Mixin to provide the public contour methods.

    Parameters
    ----------
    formatter : callable
        A conversion function to convert from the internal Matplotlib contour
        format to an external format.  See :ref:`formatters` for more
        information.

    Attributes
    ----------
    formatter : :func:`callable`
        A conversion function to convert from the internal Matplotlib contour
        format to an external format.  See :ref:`formatters` for more
        information.

    """

    __slots__ = ('formatter',)

    def __init__(self, formatter=numpy_formatter, *args, **kwargs):
        """Initialize a :class:`ContourMixin`, see class docstring."""
        # pylint: disable=unused-argument
        self.formatter = formatter

    def contour(self, level):
        """Get contour lines at the given level.

        Parameters
        ----------
        level : numbers.Number
            The data level to calculate the contour lines for.

        Returns
        -------
        :
            The result of the :attr:`formatter` called on the contour at the
            given `level`.

        """
        if not isinstance(level, numbers.Number):
            raise TypeError(
                ("'_level' must be of type 'numbers.Number' but is "
                 "'{:s}'").format(type(level)))
        vertices = self._contour_generator.create_contour(level)
        return self.formatter(level, vertices)

    def filled_contour(self, min=None, max=None):
        """Get contour polygons between the given levels.

        Parameters
        ----------
        min : numbers.Number or None
            The minimum data level of the contour polygon.  If :obj:`None`,
            ``numpy.finfo(numpy.float64).min`` will be used.
        max : numbers.Number or None
            The maximum data level of the contour polygon.  If :obj:`None`,
            ``numpy.finfo(numpy.float64).max`` will be used.

        Returns
        -------
        :
            The result of the :attr:`formatter` called on the filled contour
            between `min` and `max`.

        """
        # pylint: disable=redefined-builtin,redefined-outer-name
        # Get the contour vertices.
        if min is None:
            min = np.finfo(np.float64).min
        if max is None:
            max = np.finfo(np.float64).max
        vertices, codes = (
            self._contour_generator.create_filled_contour(min, max))
        return self.formatter((min, max), vertices, codes)


contours/quad.py
# -* coding: utf-8 -*-
"""Contour calculation of data on structured grids.

    See [1]_ for definition of structured grid types.

.. _Shapely speedups: http://toblerity.org/shapely/manual.html#performance

References
----------
.. [1] Structured grids. (n.d.). Retrieved July 21, 2017, from
   https://geo-ide.noaa.gov/wiki/index.php?title=Structured_grids

"""

# Python 2 support
# pylint: disable=redefined-builtin,unused-wildcard-import,wildcard-import
from __future__ import (absolute_import, division,
                        print_function, unicode_literals)
from builtins import *

import numpy as np
import matplotlib._contour as _contour
from .core import ContourMixin, numpy_formatter


class QuadContourGenerator(ContourMixin):
    """Contour line/polygon generator.

    The default constructor supports general structured (curvilinear) grids.
    If the grid is more structured consider using one of the other
    constructors:

    * :meth:`from_curvilinear` - same as default constructor
    * :meth:`from_rectilinear` - rectilinear grids
    * :meth:`from_uniform` - uniform or Cartesian grids

    .. _Matplotlib: https://matplotlib.org/

    Note
    ----
    This is a wrapper around Matplotlib's :class:`QuadContourGenerator
    <matplotlib.QuadContourGenerator>`, a highly efficient contour generator
    written in pure C++.

    Example
    -------
    A contrived example using :class:`QuadContourGenerator` to compute the area
    and circumference of a circle and a ring.

    >>> from contours.core import shapely_formatter as shapely_fmt
    >>> from contours.quad import QuadContourGenerator
    >>> import numpy as np
    >>> x = np.arange(-1, 1+0.01, 0.01)
    >>> y = np.arange(-1, 1+0.01, 0.01)
    >>> z = np.sqrt(x[:, np.newaxis]**2 + y[np.newaxis, :]**2)
    >>> c = QuadContourGenerator.from_rectilinear(x, y, z, shapely_fmt)
    >>> contour = c.filled_contour(max=1.0)
    >>> print('Area: {:0.2f}'.format(contour[0].area))
    Area: 3.14
    >>> print('Length: {:0.2f}'.format(contour[0].length))
    Length: 6.28
    >>> contour = c.filled_contour(min=0.5, max=1.0)
    >>> print('Area: {:0.2f}'.format(contour[0].area))
    Area: 2.36
    >>> print('Length: {:0.2f}'.format(contour[0].exterior.length))
    Length: 6.28
    >>> print('Length: {:0.2f}'.format(contour[0].interiors[0].length))
    Length: 3.14

    Parameters
    ----------
    x : array_like
        x coordinates of each point in `z`.  Must be the same size as `z`.
    y : array_like
        y coordinates of each point in `z`.  Must be the same size as `z`.
    z : array_like
        The 2-dimensional (possibly curvilinear) grid of data to compute
        contours for.  Masked arrays are supported.
    formatter : callable
        A conversion function to convert from the internal `Matplotlib`_
        contour format to an external format.  See :ref:`formatters` for more
        information.

    Attributes
    ----------
    formatter : :func:`callable`
        A conversion function to convert from the internal `Matplotlib`_
        contour format to an external format.  See :ref:`formatters` for more
        information.

    """

    slots = ('_contour_generator',)

    def __init__(self, x, y, z, formatter=numpy_formatter):
        """Initialize a :class:`QuadContourGenerator`, see class docstring."""
        x = np.asarray(x, dtype=np.float64)
        y = np.asarray(y, dtype=np.float64)
        z = np.ma.asarray(z, dtype=np.float64)
        if x.shape != z.shape:
            raise TypeError(
                ("'x' and 'z' must be the same shape but 'x' has shape {:s} "
                 "and 'z' has shape {:s}").format(str(x.shape), str(z.shape)))
        if y.shape != z.shape:
            raise TypeError(
                ("'y' and 'z' must be the same shape but 'y' has shape {:s} "
                 "and 'z' has shape {:s}").format(str(y.shape), str(z.shape)))
        mask = z.mask if z.mask is not np.ma.nomask and z.mask.any() else None
        self._contour_generator = _contour.QuadContourGenerator(
            x, y, z.filled(), mask, True, 0)
        super().__init__(formatter)

    @classmethod
    def from_curvilinear(cls, x, y, z, formatter=numpy_formatter):
        """Construct a contour generator from a curvilinear grid.

        Note
        ----
        This is an alias for the default constructor.

        Parameters
        ----------
        x : array_like
            x coordinates of each point in `z`.  Must be the same size as `z`.
        y : array_like
            y coordinates of each point in `z`.  Must be the same size as `z`.
        z : array_like
            The 2-dimensional curvilinear grid of data to compute
            contours for.  Masked arrays are supported.
        formatter : callable
            A conversion function to convert from the internal `Matplotlib`_
            contour format to an external format.  See :ref:`formatters` for
            more information.

        Returns
        -------
        : :class:`QuadContourGenerator`
            Initialized contour generator.

        """
        return cls(x, y, z, formatter)

    @classmethod
    def from_rectilinear(cls, x, y, z, formatter=numpy_formatter):
        """Construct a contour generator from a rectilinear grid.

        Parameters
        ----------
        x : array_like
            x coordinates of each column of `z`.  Must be the same length as
            the number of columns in `z`.  (len(x) == z.shape[1])
        y : array_like
            y coordinates of each row of `z`.  Must be the same length as the
            number of columns in `z`.  (len(y) == z.shape[0])
        z : array_like
            The 2-dimensional rectilinear grid of data to compute contours for.
            Masked arrays are supported.
        formatter : callable
            A conversion function to convert from the internal `Matplotlib`_
            contour format to an external format.  See :ref:`formatters` for
            more information.

        Returns
        -------
        : :class:`QuadContourGenerator`
            Initialized contour generator.

        """
        x = np.asarray(x, dtype=np.float64)
        y = np.asarray(y, dtype=np.float64)
        z = np.ma.asarray(z, dtype=np.float64)
        # Check arguments.
        if x.ndim != 1:
            raise TypeError(
                "'x' must be a 1D array but is a {:d}D array".format(x.ndim))
        if y.ndim != 1:
            raise TypeError(
                "'y' must be a 1D array but is a {:d}D array".format(y.ndim))
        if z.ndim != 2:
            raise TypeError(
                "'z' must be a 2D array but it a {:d}D array".format(z.ndim))
        if x.size != z.shape[1]:
            raise TypeError(
                ("the length of 'x' must be equal to the number of columns in "
                 "'z' but the length of 'x' is {:d} and 'z' has {:d} "
                 "columns").format(x.size, z.shape[1]))
        if y.size != z.shape[0]:
            raise TypeError(
                ("the length of 'y' must be equal to the number of rows in "
                 "'z' but the length of 'y' is {:d} and 'z' has {:d} "
                 "rows").format(y.size, z.shape[0]))
        # Convert to curvilinear format and call constructor.
        y, x = np.meshgrid(y, x, indexing='ij')
        return cls(x, y, z, formatter)

    @classmethod
    def from_uniform(
            cls, z, origin=(0, 0), step=(1, 1), formatter=numpy_formatter):
        """Construct a contour generator from a uniform grid.

        NOTE
        ----
        The default `origin` and `step` values is equivalent to calling
        :meth:`matplotlib.axes.Axes.contour` with only the `z` argument.

        Parameters
        ----------
        z : array_like
            The 2-dimensional uniform grid of data to compute contours for.
            Masked arrays are supported.
        origin : (number.Number, number.Number)
            The (x, y) coordinate of data point `z[0,0]`.
        step :  (number.Number, number.Number)
            The (x, y) distance between data points in `z`.
        formatter : callable
            A conversion function to convert from the internal `Matplotlib`_
            contour format to an external format.  See :ref:`formatters` for
            more information.

        Returns
        -------
        : :class:`QuadContourGenerator`
            Initialized contour generator.

        """
        z = np.ma.asarray(z, dtype=np.float64)
        # Check arguments.
        if z.ndim != 2:
            raise TypeError(
                "'z' must be a 2D array but it a {:d}D array".format(z.ndim))
        if len(origin) != 2:
            raise TypeError(
                "'origin' must be of length 2 but has length {:d}".format(
                    len(origin)))
        if len(step) != 2:
            raise TypeError(
                "'step' must be of length 2 but has length {:d}".format(
                    len(step)))
        if any(s == 0 for s in step):
            raise ValueError(
                "'step' must have non-zero values but is {:s}".format(
                    str(step)))
        # Convert to curvilinear format and call constructor.
        y, x = np.mgrid[
            origin[0]:(origin[0]+step[0]*z.shape[0]):step[0],
            origin[1]:(origin[1]+step[1]*z.shape[1]):step[1]]
        return cls(x, y, z, formatter)
